archive/imo1988_q6.lean

```diff
@@ -0,0 +1,282 @@
+/-
+Copyright (c) 2019 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.rat.basic
+import tactic
+
+/-!
+# IMO1988 Q6 and constant descent Vieta jumping
+
+Question 6 of IMO1988 is somewhat (in)famous. Several expert problem solvers
+could not tackle the question within the given time limit.
+The problem lead to the introduction of a new proof technique,
+so called “Vieta jumping”.
+
+In this file we formalise constant descent Vieta jumping,
+and apply this to prove Q6 of IMO1988.
+To illustrate the technique, we also prove a similar result.
+
+-/
+
+-- open_locale classical
+
+local attribute [instance] classical.prop_decidable
+local attribute [simp] nat.pow_two
+
+/-- Constant descent Vieta jumping is a proof technique
+that can typically be used to show that certain algebraic curves of degree 2
+in the affine plane do not have integral points (or only an explicit set of solutions).
+
+We need a long list of assumptions for this lemma.
+The upside is that it takes care of several annoying edge cases.
+First of all, the user should provide x and y
+that satisfy the predicate H(x,y).
+The functions B(x) and C(x) give the coefficient of the quadratic equation
+y^2 - B(x) * y + C(x) = 0.
+The predicate `base` allows the user to specify a locus of base cases
+that will be proven apart from the descent step `H_desc`.
+This is the end of the input data. What follows are proof obligations.
+In `H_quad` the user must supply a proof that H(x,y) is satisfied
+if and only if the quadratic equation y^2 - B(x) * y + C(x) = 0 is satisfied.
+Next the user should prove that H is symmetric in `H_symm`.
+Then the user should prove H(x,0) and H(x,x) for every x,
+in `H_zero` and `H_diag` respectively.
+Finally, the user should prove the descent step in `H_desc`,
+and the remaining bases cases in `H_base`. -/
+lemma constant_descent_vieta_jumping (x y : ℕ) {claim : Prop} {H : ℕ → ℕ → Prop}
+  (h₀ : H x y) (B : ℕ → ℤ) (C : ℕ → ℤ) (base : ℕ → ℕ → Prop)
+  (H_quad : ∀ {x y}, H x y ↔ (y:ℤ) * y - B x * y + C x = 0) (H_symm : ∀ {x y}, H x y ↔ H y x)
+  (H_zero : ∀ {x}, H x 0 → claim) (H_diag : ∀ {x}, H x x → claim)
+  (H_desc : ∀ {x y}, x < y → x > 0 → ¬base x y → H x y →
```

Comment:
```suggestion
  (H_desc : ∀ {x y}, 0 < x → x < y → ¬base x y → H x y →
```
if it doesn't cause too much trouble for the proof. Same on the next line (prefer le to ge).

Reply:
Fixed